aten/src/ATen/SparseTensorImpl.cpp

```diff
@@ -57,29 +66,56 @@ std::unique_ptr<Storage> SparseTensorImpl::storage() {
+// NOTE: this function is unsafe because it doesn't check whether any indices are
+// out of boundaries of `sizes`, so it should ONLY be used where we know that the
+// indices are guaranteed to be within bounds.
+#ifndef USE_TH_SIZE_ZERO_DIM
 void SparseTensorImpl::set_indices_and_values(const Tensor& indices, const Tensor& values) {
-  // TODO: Explicit empty test is needed because we don't handle size zero
-  // dimensions at the moment
-  bool empty = values.numel() == 0;
+#else
+void SparseTensorImpl::set_indices_and_values_unsafe(const Tensor& indices, const Tensor& values) {
+#endif
   AT_CHECK(values.type().toSparse() == type(), "values type must match sparse tensor type");
   AT_CHECK(indices.type().scalarType() == kLong, "indices must be an int64 tensor");
   AT_CHECK(indices.type().backend() == values.type().backend(), "backend of indices (", indices.type().backend(), ") must match backend of values (", values.type().backend(), ")");
   AT_CHECK(!indices.is_cuda() || indices.get_device() == values.get_device(), "device of indices (", indices.get_device(), ") must match device of values (", values.get_device(), ")");
+
+#ifndef USE_TH_SIZE_ZERO_DIM
+  // TODO: Explicit empty test is needed because we don't handle size zero
+  // dimensions at the moment
+  bool empty = values.numel() == 0;
   if (!empty) {
-    AT_CHECK(indices.dim() == 2, "indices must be nDim x nnz");
-    AT_CHECK(indices.size(1) == values.size(0), "indices and values must have same nnz");
+    AT_CHECK(indices.dim() == 2, "indices must be nDim x nnz, but got: ", indices.sizes());
+    AT_CHECK(indices.size(1) == values.size(0), "indices and values must have same nnz, but got nnz from indices: ", indices.size(1), ", nnz from values: ", values.size(0));
     AT_CHECK(indices.size(0) == sparseDims_, "indices has incorrect first dimension, expected ", sparseDims_, ", got ", indices.size(0));
     AT_CHECK(values.dim() == denseDims_ + 1, "values has incorrect number of dimensions, expected ", denseDims_ + 1, ", got ", values.dim());
   } else {
     AT_CHECK(indices.numel() == 0, "if values is empty, indices must be empty too");
   }
+#else
+  AT_CHECK(indices.dim() == 2, "indices must be nDim x nnz, but got: ", indices.sizes());
+  AT_CHECK(indices.size(1) == values.size(0), "indices and values must have same nnz, but got nnz from indices: ", indices.size(1), ", nnz from values: ", values.size(0));
+  AT_CHECK(indices.size(0) == sparseDims_, "indices has incorrect first dimension, expected ", sparseDims_, ", got ", indices.size(0));
+  AT_CHECK(values.dim() == denseDims_ + 1, "values has incorrect number of dimensions, expected ", denseDims_ + 1, ", got ", values.dim());
+
+  auto dense_size_original = sizes().slice(sparseDims_);
+  auto dense_size_new = values.sizes().slice(1);
+  AT_CHECK(dense_size_original.equals(dense_size_new), "values has incorrect dense dims size, expected ", dense_size_original, ", got ", dense_size_new);
```

Comment:
Wouldn't it be clearer if the error message was in terms of the entire values size (i.e. got values size, ..., expected values size, ...), instead of the the slice of the values size?  The check itself looks good, though.